### DISCUSSION TITLE
Using the AudioEffect.h  parameter "set" methods

### DISCUSSION THREAD

**Discussion by jmtalbert** [pschatzmann/arduino-audio-tools#116] [Q&A] [answered]
In the Main Loop() I would like to use one of the AudioEffect parameter "set" methods, such as setFuzzEffectValue() from the Fuzz class.  Haven't been able to hit upon the right way to build this one line.    Here is the setup and main loop of my code:

//Audio Setup
SineWaveGenerator<int16_t> sine;     
AudioEffects<SineWaveGenerator<int16_t>> effects(sine);  
GeneratedSoundStream<int16_t> in(effects);  

AudioKitStream out;                         
StreamCopy copier(out, in);             

// ~~~~~~~~~~~~~~~~~~~  SETUP  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

void setup(void) {  

  effects.addEffect(new Fuzz(fuzzEffectValue));
  
  auto config = out.defaultConfig(TX_MODE);                    
  config.sample_rate = sample_rate;
  config.channels = channels;
  config.bits_per_sample = 16;
  config.sd_active = false;
  
  out.begin(config);                                                                                 
  sine.begin(config, N_B4);                                   
  in.begin(config);  
}

/// ~~~~~~~~~~~~~~~~~~~  MAIN LOOP  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

void loop() {
  
  if(!digitalRead(KEY1)){  
    sine.setFrequency(100 + analogRead(POT1));
    //    audio_tools::Fuzz.setFuzzEffectValue(analogRead(POT2)) ; //  NOT RIGHT!!
  } 
  copier.copy();                                              
}

**Comment by pschatzmann** [comment, marked as answer]
I would expect that the following should work: You create the fuzz as global variable so that you can use it in your sketch. I assume that your pot provides values in the range from 0 to 1023, so I map them to a new value range of 0 and 20.  I am not sure what the best value range is however and I let you play with it...

```
//Audio Setup
SineWaveGenerator<int16_t> sine;
AudioEffects<SineWaveGenerator<int16_t>> effects(sine);
GeneratedSoundStream<int16_t> in(effects);
Fuzz fuzz;
AudioKitStream out;
StreamCopy copier(out, in);

// ~~~~~~~~~~~~~~~~~~~ SETUP ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

void setup(void) {

effects.addEffect(fuzz);

auto config = out.defaultConfig(TX_MODE);
config.sample_rate = sample_rate;
config.channels = channels;
config.bits_per_sample = 16;
config.sd_active = false;

out.begin(config);
sine.begin(config, N_B4);
in.begin(config);
}

/// ~~~~~~~~~~~~~~~~~~~ MAIN LOOP ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

void loop() {

if(!digitalRead(KEY1)){
sine.setFrequency(100 + analogRead(POT1));
int effectValue = map(analogRead(POT2),0, 1023, 0, 20);
fuzz.setFuzzEffectValue(effectValue) ;
}
copier.copy();
}
```

**Comment by jmtalbert** [comment]
Thanks so much, that worked great!   I've attached the final sketch if anyone is interested.
[ES8388_Sine_Distortion.txt](https://github.com/pschatzmann/arduino-audio-tools/files/8193234/ES8388_Sine_Distortion.txt)

**Comment by pschatzmann** [comment]
Did you try with multiple effects in a chain ? 
If you have separate pots you can control them individually...

**Comment by jmtalbert** [comment]
Yes, I got several effects working in a chain with real-time effect parameter adjustments from several attached pots.  Since the ESP32 has two processing cores, I assigned the task of updating effect settings to the second core and left the main loop to deal only with copier().  Works like a charm with no extraneous noise.  I'll attach my script.  

The next step is to use the ES8388 input instead of a sinewave.  This would be AudioKit to Effects to AudioKit.  I'm having trouble setting this up...

AudioKitStream kit;                                         
GeneratedSoundStream<int16_t> sound(kit);  //This line won't compile
AudioEffects<int16_t>   effects(sound);
GeneratedSoundStream<int16_t> in(effects);
Fuzz fuzz(fuzzEffectValue); 
StreamCopy copier(in, kit);   

[ES8388_Sine_Effects_2Core.txt](https://github.com/pschatzmann/arduino-audio-tools/files/8202185/ES8388_Sine_Effects_2Core.txt)

**Comment by pschatzmann** [comment]
The GeneratedSoundStream expects a SoundGenerator as input but AudioKitStream implements the Stream interface. 
You can use the class GeneratorFromStream() to convert a Stream to a Generator...

ps. I just committed a change that let you pass a stream to the AudioEffects:

```
        AudioEffects(Stream &input, int channels=2, float volume=1.0) 
```
So the following should compile now:
```
AudioKitStream kit;
AudioEffects<GeneratorFromStream<effect_t>> effects(kit,2);
GeneratedSoundStream<int16_t> in(effects);
```
Just a small disclaimer. I did not have the time to test this properly...

**Comment by jmtalbert** [comment]
Thanks so much for your Audio Libraries and your help.  I finished writing
up my project.  It includes 5 Arduino sketches based on your Libraries.
You are welcome to use any of it if you think it is helpful.

Check out the downloadable PDF on the "ESP32 ES8388 Codec" project, bottom
of page at https://jtalbert.xyz/ESP32/

I was especially interested in your Effects.  Your Delay Effect didn't seem
to do anything.  I'm hoping to study that one and try expanding your
Effects Library with  some other effects on live signals such as modulation
and granular synthesis.

John Talbert

**Comment by pschatzmann** [comment]
Thanks a lot for your feed back. I suggest that you add your project to https://github.com/pschatzmann/arduino-audio-tools/discussions/categories/show-and-tell

If you are interrested in effects you might also have a look at Maximilian...
